> These functions were only part of the API if perl is built with clang.
> However perl modules built with clang still try to use them even
> when perl itself is built with gcc. This patch removes the #ifdef PERL_TSA_ACTIVE
> around these functions so they are always available and fixes the build of
> perl modules with clang when perl is built with gcc.

This fails very quickly during `make` when building on Linux with `gcc`:
```
cc -c -DPERL_CORE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement reentr.c
util.c:6558:25: error: unknown type name ‘perl_mutex’; did you mean ‘Perl_pp_hex’?
 int perl_tsa_mutex_lock(perl_mutex* mutex)
                         ^~~~~~~~~~
                         Perl_pp_hex
util.c:6563:27: error: unknown type name ‘perl_mutex’; did you mean ‘Perl_pp_hex’?
 int perl_tsa_mutex_unlock(perl_mutex* mutex)
                           ^~~~~~~~~~
                           Perl_pp_hex
util.c:6568:28: error: unknown type name ‘perl_mutex’; did you mean ‘Perl_pp_hex’?
 int perl_tsa_mutex_destroy(perl_mutex* mutex)
                            ^~~~~~~~~~
                            Perl_pp_hex
makefile:250: recipe for target 'util.o' failed
make: *** [util.o] Error 1
make: *** Waiting for unfinished jobs....
```
Or see any of the github.actions listed in this p.r., e.g., https://github.com/Perl/perl5/pull/17786/checks?check_run_id=678552324

It also fails with `clang` on Linux, albeit with somewhat different error messages:
```
clang -c -DPERL_CORE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement util.c
util.c:6558:25: error: unknown type name 'perl_mutex'
int perl_tsa_mutex_lock(perl_mutex* mutex)
                        ^
util.c:6560:12: warning: implicit declaration of function 'pthread_mutex_lock'
      [-Wimplicit-function-declaration]
    return pthread_mutex_lock((pthread_mutex_t *) mutex);
           ^
util.c:6563:27: error: unknown type name 'perl_mutex'
int perl_tsa_mutex_unlock(perl_mutex* mutex)
                          ^
util.c:6565:12: warning: implicit declaration of function 'pthread_mutex_unlock'
      [-Wimplicit-function-declaration]
    return pthread_mutex_unlock((pthread_mutex_t *) mutex);
           ^
util.c:6568:28: error: unknown type name 'perl_mutex'
int perl_tsa_mutex_destroy(perl_mutex* mutex)
                           ^
util.c:6570:12: warning: implicit declaration of function 'pthread_mutex_destroy'
      [-Wimplicit-function-declaration]
    return pthread_mutex_destroy((pthread_mutex_t *) mutex);
           ^
3 warnings and 3 errors generated.
makefile:250: recipe for target 'util.o' failed
make: *** [util.o] Error 1
```

Also, I do not understand what you mean when you refer to a perl module "built" with one compiler against a `perl` build with a different compiler.  Can you clarify?

As it stands, I do not think this pull request should be applied.

Thank you very much.
Jim Keenan

--- a/AUTHORS
+++ b/AUTHORS
@@ -1257,6 +1257,7 @@ Tom Hughes                     <tom@compton.nu>
 Tom Hukins                     <tom@eborcom.com>
 Tom Phoenix                    <rootbeer@teleport.com>
 Tom Spindler                   <dogcow@isi.net>
+Tom Stellard                   <tstellar@redhat.com>
 Tom Wyant                      <wyant@cpan.org>
 Tomasz Konojacki               <me@xenu.pl>
 Tomoyuki Sadahiro              <BQW10602@nifty.com>

--- a/util.c
+++ b/util.c
@@ -6552,7 +6552,7 @@ Perl_dump_c_backtrace(pTHX_ PerlIO* fp, int depth, int skip)
 
 #endif /* #ifdef USE_C_BACKTRACE */
 
-#ifdef PERL_TSA_ACTIVE
+#if defined(USE_ITHREADS) && defined(I_PTHREAD)
 
 /* pthread_mutex_t and perl_mutex are typedef equivalent
  * so casting the pointers is fine. */
@@ -6573,7 +6573,6 @@ int perl_tsa_mutex_destroy(perl_mutex* mutex)
 }
 
 #endif
-
 
 #ifdef USE_DTRACE
 

